data/default.nix

```diff
@@ -1,42 +1,54 @@
 { nixpkgs ? <nixpkgs>
 , config ? {}
 }:
-
 with (import nixpkgs config);
-
 let
   mkDerivation =
-    { srcs ? ./elm-srcs.nix
-    , src
+    { elmsrcs ? ./elm-srcs.nix
+    , srcs
+    , srcdir ? "src"
     , name
-    , srcdir ? "./src"
     , targets ? []
     , versionsDat ? ./versions.dat
     }:
-    stdenv.mkDerivation {
-      inherit name src;
+    let sanitizePath = str:
+        let path = builtins.filter (p: p != "..") (lib.splitString "/" str);
+        in lib.concatStringsSep "/" (map (p: if p == "." then srcdir else p) path);
+    in stdenv.mkDerivation {
+      inherit name srcs;
+      sourceRoot = ".";
 
       buildInputs = [ elmPackages.elm ];
 
+      patchPhase = let
+        elmjson = let json = (lib.importJSON ./elm.json) // { source-directories = map sanitizePath srcs; };
+        in writeText "elm.json" (builtins.toJSON json);
+      in ''
+        echo "Generating elm.json"
+        cp \${elmjson} ./elm.json
+        cat elm.json
+      '';
+
       buildPhase = pkgs.elmPackages.fetchElmDeps {
-        elmPackages = import srcs;
+        elmPackages = import elmsrcs;
         inherit versionsDat;
       };
 
-      installPhase = let
-        elmfile = module: "\${srcdir}/\${builtins.replaceStrings ["."] ["/"] module}.elm";
-      in ''
-        mkdir -p \$out/share/doc
-        \${lib.concatStrings (map (module: ''
-          echo "compiling \${elmfile module}"
-          elm make \${elmfile module} --output \$out/\${module}.html --docs \$out/share/doc/\${module}.json
-        '') targets)}
+      installPhase = ''
+        mkdir -p $out/share/doc
+
+        \${lib.concatStrings (map (module:
+          let fullmodule = sanitizePath module;
+              modulename = sanitizePath (lib.removePrefix "./" module);
+          in ''
+            echo "compiling \${module}"
+            elm make \${fullmodule}.elm --output $out/\${modulename}.html --docs $out/share/doc/\${modulename}.json
+          '') targets)}
       '';
     };
 in mkDerivation {
   name = "${name}";
-  srcs = ./elm-srcs.nix;
-  src = ./.;
-  targets = ["Main"];
+  srcs = ${srcs};
   srcdir = "${srcdir}";
+  targets = ["./Main"];
```

Comment:
BREAKING change: `Main.Entry` => `./Main.Entry`. Supports custom extensions like `Main.js.elm` or `Main.html.elm` and entries from different sources (relative path from current dir).

Reply:
Can we revert this then?

Reply:
unfortunetely this is still required.

Imagine you have `source-directories` like this defined in `elm.json`.

```
"source-directories": [
  "./src",
  "./src2",
  "../lib"
]
```

Now when you set `targets = [ "Main" ]` it's not clear in which directory this main should be located. It can be any one of those.

Another issue is that with current dir. Imagine having this `elm.json`:

```
"source-directories": [
   ".",
   "../lib"
]
```

since I've used `srcs` attribute in derivation nix will create structure as:

- $pwd # contains files form `.`, is named after current durectory
- lib # contains files from `../lib`

this is why Main is prefixed with `./`. Thanks to this we know that it should be what ever current directory was named after. This what `srcdir` variable is used for.

I wonder if there is any buildin helper that would help to resolve paths in a way nix does it when `srcs` on mkDerivation is used. I belive this is  done in a [this shellscript](https://github.com/NixOS/nixpkgs/blob/076860e0340a5e4a909b9a710e186508b14d1c90/pkgs/stdenv/generic/setup.sh#L893-L896)